akka-projection-r2dbc/src/it/scala/akka/projection/r2dbc/ChangeSliceRangesSpec.scala

```diff
@@ -0,0 +1,256 @@
+
+  val config: Config = ConfigFactory
+    .parseString("""
+    akka.persistence.r2dbc {
+      query {
+        backtracking {
+          window = 5 seconds
+          behind-current-time = 3 seconds
+        }
+      }
+    }
+    """)
+    .withFallback(TestConfig.config)
+
+  final case class Processed(projectionId: ProjectionId, envelope: EventEnvelope[String])
+
+}
+
+class ChangeSliceRangesSpec
+    extends ScalaTestWithActorTestKit(ChangeSliceRangesSpec.config)
+    with AnyWordSpecLike
+    with TestDbLifecycle
+    with TestData
+    with LogCapturing {
+  import ChangeSliceRangesSpec._
+  import EventSourcedEndToEndSpec.Persister
+
+  override def typedSystem: ActorSystem[_] = system
+
+  private val log = LoggerFactory.getLogger(getClass)
+
+  private val projectionSettings = R2dbcProjectionSettings(system)
+
+  private class TestHandler(projectionId: ProjectionId, probe: ActorRef[Processed], delaySlice: Int)
+      extends R2dbcHandler[EventEnvelope[String]] {
+    private val log = LoggerFactory.getLogger(getClass)
+
+    override def process(session: R2dbcSession, envelope: EventEnvelope[String]): Future[Done] = {
+      val slice = persistenceExt.sliceForPersistenceId(envelope.persistenceId)
+      log.debugN("{} Processed {}, pid {}, slice {}", projectionId.key, envelope.event, envelope.persistenceId, slice)
+      probe ! Processed(projectionId, envelope)
+      if (slice == delaySlice)
+        akka.pattern.after(3.second)(Future.successful(Done))
+      else
+        Future.successful(Done)
+    }
+  }
+
+  private def startProjections(
+      entityType: String,
+      projectionName: String,
+      nrOfProjections: Int,
+      processedProbe: ActorRef[Processed],
+      delaySlice: Int = -1): Vector[ActorRef[ProjectionBehavior.Command]] = {
+    val sliceRanges = EventSourcedProvider.sliceRanges(system, R2dbcReadJournal.Identifier, nrOfProjections)
+
+    sliceRanges.map { range =>
+      val projectionId = ProjectionId(projectionName, s"${range.min}-${range.max}")
+      val sourceProvider =
+        EventSourcedProvider
+          .eventsBySlices[String](system, R2dbcReadJournal.Identifier, entityType, range.min, range.max)
+      val projection = R2dbcProjection
+        .exactlyOnce(
+          projectionId,
+          Some(projectionSettings),
+          sourceProvider = sourceProvider,
+          handler = () => new TestHandler(projectionId, processedProbe.ref, delaySlice))
+      spawn(ProjectionBehavior(projection))
+    }.toVector
+  }
+
+  def persistenceIdForSlice(entityType: String, slice: Int): PersistenceId = {
+    @tailrec def loop(n: Int): PersistenceId = {
+      val candidate = PersistenceId(entityType, s"p$n")
+      if (persistenceExt.sliceForPersistenceId(candidate.id) == slice)
+        candidate
+      else
+        loop(n + 1)
+    }
+    loop(0)
+  }
+
+  private def mkEvent(n: Int): String = {
+    val template = "0000000"
+    val s = n.toString
+    "e" + (template + s).takeRight(5)
+  }
```

Comment:
Not important, but could be format string:

```suggestion
  private def mkEvent(n: Int): String = f"e$n%05d"
```